### PR TITLE
Avoid duplicate OBJECT_IDs after MVU

### PR DESCRIPTION
**Pull request by 2jungkook**
Signed-off-by: Jungkook Lee <jungkook@amazon.com>

### Description

PG allows duplicated OIDs in different catalog tables. However, in Babelfish, we use PG's OID as OBJECT_ID in all system catalog views and it should be unique within a logical database (https://github.com/MicrosoftDocs/sql-docs/blob/live/docs/relational-databases/system-catalog-views/sys-all-objects-transact-sql.md). Duplicate OIDs can break SQL server tools, e.g., SSMS, as well as customer applications.

In PG/Babelfish, duplicate OIDs can occur in the following two cases:
1. After MVU (Major Version Upgrade)
2. After hitting OID wraparound

This commit is to prevent duplicate OIDs during MVU and consists of two parts:

> First of all, from the old cluster, it finds the max OID in the 5 catalog tables: `pg_extension`, `pg_authid`, `pg_enum`, `pg_class` and `pg_type`. And then pass the max OID to the new cluster as a form of GUC, `babelfishpg_tsql.dump_restore_min_oid`. In the new cluster, set the nextOid greater than the max OID from the old cluster using `GetNewObjectId_hook`.

> Secondly, in `pg_upgrade.c`, it skips the step of transferring the next OID from the old cluster to the new cluster.

> Also note that the type of babelfishpg_tsql.dump_restore_min_oid is defined by `DefineCustomStringVariable()` to support UINT_MAX.

Extension PR: https://github.com/babelfish-for-postgresql/babelfish_extensions/pull/693
 
### Issues Resolved

BABEL-3613

### Test Scenarios Covered ###
* **Use case based -** Yes
* **Boundary conditions -** Yes (babelfishpg_tsql.dump_restore_min_oid)
* **Arbitrary inputs -** N/A
* **Negative test cases -** N/A
* **Minor version upgrade tests -** Yes
* **Major version upgrade tests -** Yes
* **Performance tests -**N/A
* **Tooling impact -** N/A
* **Client tests -** N/A
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/access/transam/varsup.c [hunk 1]
```diff
@@ -569,6 +570,9 @@ GetNewObjectId(void)
 		}
 	}
 
+	if (GetNewObjectId_hook)
```
Comment (surendravishnoi) [inline src/backend/access/transam/varsup.c:573]: this will add cost for each get oid under lock. Can we do it one time before any object is restored? may be when tsql extension is loaded?

Comment (2jungkook) [reply src/backend/access/transam/varsup.c:573]: Eventually, we will need a hook when we make a full fix including the wraparound case.

File: src/bin/pg_upgrade/pg_upgrade.c [hunk 2]
```diff
@@ -162,12 +166,19 @@ main(int argc, char **argv)
...
-	prep_status("Setting next OID for new cluster");
-	exec_prog(UTILITY_LOG_FILE, NULL, true, true,
-			  "\"%s/pg_resetwal\" -o %u \"%s\"",
-			  new_cluster.bindir, old_cluster.controldata.chkpnt_nxtoid,
-			  new_cluster.pgdata);
-	check_ok();
+	if (!has_bbf_db)
+	{
+		/*
+		 * In Babelfish, if we transfer the old cluster's next OID to the new cluster,
```
Comment (surendravishnoi) [inline src/bin/pg_upgrade/pg_upgrade.c]: this comment should got outside and existing comment should come inside

File: src/bin/pg_upgrade/pg_upgrade.c [hunk 3]
```diff
@@ -162,12 +166,19 @@ main(int argc, char **argv)
...
+		 * it can make a duplicate OID that breaks a Babelfish's assumption.
+		 */
+		prep_status("Setting next OID for new cluster");
+		exec_prog(UTILITY_LOG_FILE, NULL, true, true,
+				"\"%s/pg_resetwal\" -o %u \"%s\"",
+				new_cluster.bindir, old_cluster.controldata.chkpnt_nxtoid,
```
Comment (surendravishnoi) [inline src/bin/pg_upgrade/pg_upgrade.c:178]: Please validate that this command is only changing oid and nothing more

File: src/bin/pg_upgrade/pg_upgrade.c [hunk 4]
```diff
@@ -558,6 +569,41 @@ copy_xact_xlog_xid(void)
...
+
+		if (ntups != 0)
+		{
+			prep_status("There is a Babelfish Database in %s\n", db_name);
+			check_ok();
+			return true;
+		}
+	}
+
+	prep_status("There is no Babelfish Database\n");
```
Comment (surendravishnoi) [inline src/bin/pg_upgrade/pg_upgrade.c]: Please remove this, it will come for APG as well

File: src/bin/pg_upgrade/pg_upgrade.c [hunk 5]
```diff
@@ -558,6 +569,41 @@ copy_xact_xlog_xid(void)
...
+	PGconn *conn;
+	int ntups;
+
+	for (int dbnum = 0; dbnum < cluster->dbarr.ndbs; dbnum++)
+	{
+		PGresult *res;
+
+		db = &cluster->dbarr.dbs[dbnum];
+		db_name = db->db_name;
+		prep_status("Checking %s has a Babelfish Database\n", db_name);
```
Comment (surendravishnoi) [inline src/bin/pg_upgrade/pg_upgrade.c]: I think we can remove this as well

Comment (2jungkook) [reply src/bin/pg_upgrade/pg_upgrade.c]: Let me move this out of the for-loop so that it only prints once.

**Review by rishabhtanwar29 — APPROVED** [review]

**Review by kuntalghosh — APPROVED** [review]
